lib/etcconf/hosts.go

```diff
@@ -163,31 +208,110 @@ func (h *hosts) SetHost(hostname string, ip net.IP) {
 }
 
 func (h *hosts) setHost(hostname string, ip net.IP) {
-	h.hosts[hostname] = ip
 	h.dirty = true
+
+	if ip == nil {
+		return
+	}
+
+	// what type of address is it?
+	hostmap := h.hostsIPv6
+	ipv4 := ip.To4()
+	if ipv4 != nil {
+		// this drops any leading garbage in the array so that byte comparisons work as
+		// expected
+		ip = ipv4
+		hostmap = h.hostsIPv4
+	}
+
+	hentry := hostmap[hostname]
+	ientry := h.entries[ip.String()]
+
+	if hentry != nil {
+		// existing entry with no changes
+		if bytes.Equal(hentry.IP, ip) {
+			h.dirty = false
+			return
+		}
+
+		// existing hostname with a new address - change address for all assocated hostnames
+		hentry.setAddress(ip)
```

Comment:
Why not just remove the host from the entry instead since it is associated with a different IP?

Reply:
because someone may have entered an extra host alias for this entry.
For example if you have the following entry:
```
10.118.20.236 container-network.localnet my-custom-alias-for-app
```
and the DHCP lease changes, then you can either:
a. edit the IP address (what this change does)
b. create a new entry and move _some_ of the hosts

So our choices are:
```
# edit the IP
10.118.22.33 container-network.localnet my-custom-alias-for-app
```
or
```
# split the entry
10.118.20.236 my-custom-alias-for-app
10.118.22.33 container-network.localnet
```

In the latter case we've likely broken some assumptions and quite possibly are still referencing an address that no longer belongs to that container.